_posts/2018-12-25-Building-A-Continues-Data-Processing-System.md

```diff
@@ -0,0 +1,89 @@
+---
+layout: post
+title:  "Building a continues data processing system"
+date:   2018-12-25 23:00:00 +0800
+categories: [blog]
+tags: [elixir, backend, back-pressure]
+author: Jun Lin
+---
+
+## A brief introduction to Archiver
+
+At RingCentral, we have a system called **Archiver**, which is a system automatically uploads customer’s call recordings, SMS and fax to their cloud storages.
+
+For example, a company connect it’s Dropbox account with Archiver, then whenever it’s employees make phone calls and recording the audio, Archiver will upload the call logs to the customer’s Dropbox storage.
+
+It seems like a really simple system, but it is still challenging to *doing great*. The heavy lifting of Archiver is it’s background workers, as some of our customers are large companies like an airline company, they may generate many call recordings in a day.
+
+To make sure the customer’s call recordings can be uploaded in time, our system starts an archiving job every 5 minutes. The job will first iterate all customers to find out how many items is pending to upload, then bootstrap a thread per customer to upload the call logs. For each customer, we then start a new thread per item to upload it to the Cloud Storage. It’s running pretty well, but as RingCentral are rapidly grows, we have encountered some issues.
+
+## What’s the problem
+
+Recently, we found many HTTP 429 errors in our logs and the whole data processing rate is slowing down. The root cause is a customer with thousands of employees has connected Archiver to their Dropbox account, and they have around hundred of thousands calls a day, so, with the Scheduler running every 5 minutes, we may hit the rate limit of Dropbox.
+
+So, to make sure the customer’s recordings will be uploaded, we have to estimate the amount of data we can process in one batch. It's affected by how frequently we start a job, how much data we need to fetch in a batch, how many threads we can run at the same time, and also the cloud storage’s rate limiting. We can eventually set a reasonable number by doing some calculation to ensure customer’s data will be uploaded eventually, but there are the complexity of the network, database IO and other bottlenecks, so it’s still an urge to find out a better solution.
```

Comment:
so it's still an urge to find out sounds quite weird. either change it to "so there is still an urge to", which is still a bit wierd, or "so it is still important to".